core/arch/arm/kernel/thread_a32.S

```diff
@@ -644,59 +658,111 @@ UNWIND(	.fnend)
 END_FUNC thread_unwind_user_mode
 
 LOCAL_FUNC thread_abort_handler , :
-thread_abort_handler:
 thread_und_handler:
 UNWIND(	.fnstart)
 UNWIND(	.cantunwind)
-	/*
-	 * Switch to abort mode to use that stack instead.
-	 */
-	cps	#CPSR_MODE_ABT
-	push	{r0-r11, ip}
-	cps	#CPSR_MODE_UND
-	mrs	r0, spsr
-	tst	r0, #CPSR_T
-	subne	r1, lr, #2
-	subeq	r1, lr, #4
-	cps	#CPSR_MODE_ABT
-	push	{r0, r1}
-	msr	spsr_fsxc, r0	/* In case some code reads spsr directly */
+	strd	r0, r1, [sp, #THREAD_CORE_LOCAL_R0]
+	mrs	r1, spsr
+	tst	r1, #CPSR_T
+	subne	lr, lr, #2
+	subeq	lr, lr, #4
 	mov	r0, #ABORT_TYPE_UNDEF
 	b	.thread_abort_generic
 
 thread_dabort_handler:
-	push	{r0-r11, ip}
-	sub	r1, lr, #8
-	mrs	r0, spsr
-	push	{r0, r1}
+	strd	r0, r1, [sp, #THREAD_CORE_LOCAL_R0]
+	sub	lr, lr, #8
 	mov	r0, #ABORT_TYPE_DATA
 	b	.thread_abort_generic
 
 thread_pabort_handler:
-	push	{r0-r11, ip}
-	sub	r1, lr, #4
-	mrs	r0, spsr
-	push	{r0, r1}
+	strd	r0, r1, [sp, #THREAD_CORE_LOCAL_R0]
+	sub	lr, lr, #4
 	mov	r0, #ABORT_TYPE_PREFETCH
-	b	.thread_abort_generic
 
 .thread_abort_generic:
+	/*
+	 * At this label:
+	 * cpsr is in mode undef or abort
+	 * sp is still pointing to struct thread_core_local belonging to
+	 * this core.
+	 * {r0-r3} are saved in struct thread_core_local pointed to by sp
+	 * {r4-r11, ip} are untouched.
+	 * r0 holds the first argument for abort_handler()
+	 */
+
+	/*
+	 * Update core local flags.
+	 * flags = (flags << THREAD_CLF_SAVED_SHIFT) | THREAD_CLF_ABORT;
+	 */
+	ldr	r1, [sp, #THREAD_CORE_LOCAL_FLAGS]
+	lsl	r1, r1, #THREAD_CLF_SAVED_SHIFT
+	orr	r1, r1, #THREAD_CLF_ABORT
+
+	/*
+	 * Select stack and update flags accordingly
+	 *
+	 * Normal case:
+	 * If the abort stack is unused select that.
+	 *
+	 * Fatal error handling:
+	 * If we're already using the abort stack as noted by bit
+	 * (THREAD_CLF_SAVED_SHIFT + THREAD_CLF_ABORT_SHIFT) in the flags
+	 * field we're selecting the temporary stack instead to be able to
+	 * make a stack trace of the abort in abort mode.
+	 *
+	 * r1 is initialized as a temporary stack pointer until we've
+	 * switched to system mode.
+	 */
+	tst	r1, #(THREAD_CLF_SAVED_SHIFT + THREAD_CLF_ABORT_SHIFT)
+	orrne	r1, r1, #THREAD_CLF_TMP /* flags |= THREAD_CLF_TMP; */
```

Comment:
True, but this is the same for AArch64. This flag is kept to be able to tell which state we're executing in. AArch64 doesn't have as many mode bits as AArch32 in CPSR so it's mostly to compensate for that.

It's probably only useful for debug purposes, but I'd prefer to keep it reliable in either case for the time being.